src/target/riscv/riscv.c

```diff
@@ -3080,6 +3123,67 @@ COMMAND_HANDLER(riscv_set_ebreaku)
 	return ERROR_OK;
 }
 
+COMMAND_HANDLER(riscv_itrigger)
+{
+	if (CMD_ARGC < 1) {
+		LOG_ERROR("Command takes at least 1 parameter");
+		return ERROR_COMMAND_SYNTAX_ERROR;
+	}
+
+	struct target *target = get_current_target(CMD_CTX);
+
+	if (riscv_enumerate_triggers(target) != ERROR_OK)
+		return ERROR_FAIL;
+
+	if (!strcmp(CMD_ARGV[0], "set")) {
+		if (find_trigger_by_id(target, -4) >= 0) {
+			LOG_ERROR("An itrigger is already set, and OpenOCD "
+				  "doesn't support setting more than one at a time.");
+			return ERROR_FAIL;
+		}
+		bool vs = false;
+		bool vu = false;
+		bool nmi = false;
+		bool m = false;
+		bool s = false;
+		bool u = false;
+		riscv_reg_t interrupts = 0;
+
+		for (unsigned int i = 1; i < CMD_ARGC; i++) {
+			if (!strcmp(CMD_ARGV[i], "vs"))
+				vs = true;
+			else if (!strcmp(CMD_ARGV[i], "vu"))
+				vu = true;
+			else if (!strcmp(CMD_ARGV[i], "nmi"))
+				nmi = true;
+			else if (!strcmp(CMD_ARGV[i], "m"))
+				m = true;
+			else if (!strcmp(CMD_ARGV[i], "s"))
+				s = true;
+			else if (!strcmp(CMD_ARGV[i], "u"))
+				u = true;
+			else
+				COMMAND_PARSE_NUMBER(u64, CMD_ARGV[i], interrupts);
+		}
+		int result = maybe_add_trigger_t4(target, vs, vu, nmi, m, s, u, interrupts, -4);
+		if (result != ERROR_OK)
+			LOG_ERROR("Failed to set requested itrigger.");
+		return result;
+
+	} else if (!strcmp(CMD_ARGV[0], "clear")) {
+		if (CMD_ARGC != 1) {
+			LOG_ERROR("clear command takes no extra arguments.");
+			return ERROR_COMMAND_SYNTAX_ERROR;
+		}
+		return remove_trigger(target, -4);
```

Comment:
I'd recommend a defined constant instead of `-4` everywhere.

For example:

```suggestion
		return remove_trigger(target, ITRIGGER_UNIQUE_ID);
```

Reply:
Done.

Reply:
I'm sorry to open this one again. This is a minor thing, but I believe important for readability:

Here I'd still recommend to have the unique ID for itrigger as a global constant.

```
// somewhere near the top of riscv.c:
#define ITRIGGER_UNIQUE_ID (-CSR_TDATA1_TYPE_ITRIGGER)

// ...

// here in COMMAND_HANDLER(riscv_itrigger):
find_first_trigger_by_id(target, ITRIGGER_UNIQUE_ID)  // etc.
```

I understand that the unique ID is not used anywhere else outside this function (yet?). But I believe it will be quicker to the reader to understand the code when the uppercase constant is there, contrary to unique_id (which looks like an ordinary variable when reading the code).

Reply:
Another possibility: Leave the code as is, just rename the variable to make it clear it is a constant.

```
const int TRIGGER_UNIQUE_ID = -CSR_TDATA1_TYPE_ITRIGGER;
```

Reply:
I went with the 2nd option.